enarx-keep-sev/src/vm.rs

```diff
@@ -0,0 +1,70 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::x86_64::*;
+
+use std::error::Error;
+
+use kvm_bindings::kvm_userspace_memory_region as MemoryRegion;
+use kvm_ioctls::{Kvm, VmFd};
+use memmap::*;
+use x86_64::structures::paging::page_table::PageTableFlags;
+use x86_64::PhysAddr;
+
+const DEFAULT_MEM_SIZE: usize = units::bytes!(1; GiB);
+
+pub struct VirtualMachine {
+    kvm: Kvm,
+    fd: VmFd,
+    address_space: Vec<MemoryRegion>,
+    mmap_handles: Vec<MmapMut>,
+}
+
+impl VirtualMachine {
+    pub fn new() -> Result<Self, Box<dyn Error>> {
+        // Create a KVM context
+        let kvm = Kvm::new()?;
+        let fd = kvm.create_vm()?;
+
+        // Create the guest address space
+        let guest_addr_space = memmap::MmapOptions::new()
+            .len(DEFAULT_MEM_SIZE)
+            .map_anon()?;
+        let region = MemoryRegion {
+            slot: 0,
+            flags: 0,
+            guest_phys_addr: 0,
+            memory_size: DEFAULT_MEM_SIZE as _,
+            userspace_addr: &*guest_addr_space as *const _ as *const u8 as u64,
+        };
```

Comment:
I don't see any flags in this allocation asking for huge pages. I'm not sure if that is necessary or not.

Reply:
The [guest's page tables enable huge pages for the guest](https://github.com/enarx/enarx/pull/536/files#diff-fd0caaaee80e2f801bb7595524d12efbR54). The keep installs those page tables into the guest's address space.

AIUI, the host's page topology is largely transparent here and I don't think it impacts the construction of our virtual machine nor should it color how the virtual machine uses the pages.

Reply:
So, the page isn't backed by an actual hugepage on the host but from what I understand, you just want the reduced complexity of setting up the page tables in the guest ?

Reply:
I suspect the Linux kernel is likely going to use a huge page implicitly for any large-enough allocation that is a multiple of the huge page size. So I'm not sure if it is strictly necessary to have the flag. Nevertheless, there are several notes buried in KVM about having huge pages on both the guest and the host have performance benefits. So it is probably worth requiring huge pages from the kernel.

Reply:
> So, the page isn't backed by an actual hugepage on the host but from what I understand, you just want the reduced complexity of setting up the page tables in the guest ?

Yes, that and we want the systems-level code to be as tiny of a footprint in the guest address space as possible

> Nevertheless, there are several notes buried in KVM about having huge pages on both the guest and the host have performance benefits. So it is probably worth requiring huge pages from the host kernel.

Ah I see I did not know that. I'll research this and follow up on this in my next revision.

Reply:
KVM does EPT mappings lazily - so with a huge page, you minimize the number of EPT/NPT exits required. And, now you also need lesser number of TLB flushes since it maps to a bigger page size but you already get a little bit of this benefit even if only the guest knows about huge pages.